DirectProgramming/DPC++FPGA/Tutorials/Features/ac_types/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,350 @@
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<Add>([=] { res[0] = x[0] + y[0]; });
+  });
+  q.wait();
+}
+
+void TestDiv(queue &q, const ac_int14 &a, const ac_int14 &b, ac_int15 &c) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_int14, 1> inp2(&b, 1);
+  buffer<ac_int15, 1> result(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<Div>([=] { res[0] = x[0] / y[0]; });
+  });
+  q.wait();
+}
+
+void TestMult(queue &q, const ac_int14 &a, const ac_int14 &b, ac_int28 &c) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_int14, 1> inp2(&b, 1);
+  buffer<ac_int28, 1> result(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<Mult>([=] { res[0] = x[0] * y[0]; });
+  });
+  q.wait();
+}
+
+void TestShiftLeft(queue &q, const ac_int14 &a, const ac_int14 &b,
+                   ac_int14 &c) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_int14, 1> inp2(&b, 1);
+  buffer<ac_int14, 1> result(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<ShiftLeft>([=] { res[0] = x[0] << y[0]; });
+  });
+  q.wait();
+}
+
+// Note how the shift amount is specified with a smaller ac_int than in
+// TestShiftLeft above
+void TestEfficientShiftLeft(queue &q, const ac_int14 &a, const ac_uint4 &b,
+                            ac_int14 &c) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_uint4, 1> inp2(&b, 1);
+  buffer<ac_int14, 1> result(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<EfficientShiftLeft>([=] { res[0] = x[0] << y[0]; });
+  });
+  q.wait();
+}
+
+// The method
+//      x = y.slc<M>(n)
+// is equivalent to the VHDL behavior of
+//      x := y((M+n-1) downto n);
+// Note that only static bit widths are supported
+void TestGetBitSlice(queue &q, const ac_int14 &a, ac_uint4 &b, const int lsb) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<int, 1> inp2(&lsb, 1);
+  buffer<ac_uint4, 1> result(&b, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<GetBitSlice>([=] { res[0] = x[0].slc<4>(y[0]); });
+  });
+  q.wait();
+}
+
+// There is a set_slc(int lsb, const ac_int<W, S> &slc) which allows the user to
+// set a bit slice as shown in the example below.
+void TestSetBitSlice(queue &q, ac_int14 &a, const ac_int4 &b, int lsb) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_int4, 1> inp2(&b, 1);
+  buffer<int, 1> inp3(&lsb, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_write};
```

Comment:
why read_write?

Reply:
Perhaps what you'd like me to do is to rename the buffer from something other than `inp` since `inp` and `read_write` together is confusing. I'll make that change.

Reply:
Do you read from it? If note, write_only?

Reply:
You are write, corrected.